src/kubernetes/state/instrumenting.rs

```diff
@@ -0,0 +1,343 @@
+    use super::*;
+    use crate::test_util;
+    use futures::{channel::mpsc, SinkExt, StreamExt};
+    use k8s_openapi::{api::core::v1::Pod, apimachinery::pkg::apis::meta::v1::ObjectMeta};
+    use once_cell::sync::OnceCell;
+    use std::sync::{Mutex, MutexGuard};
+
+    fn prepare_test() -> (
+        Writer<mock::Writer<Pod>>,
+        mpsc::Receiver<mock::ScenarioEvent<Pod>>,
+        mpsc::Sender<()>,
+    ) {
+        let (events_tx, events_rx) = mpsc::channel(0);
+        let (actions_tx, actions_rx) = mpsc::channel(0);
+        let writer = mock::Writer::new(events_tx, actions_rx);
+        let writer = Writer::new(writer);
+        (writer, events_rx, actions_tx)
+    }
+
+    fn make_pod() -> Pod {
+        Pod {
+            metadata: Some(ObjectMeta {
+                name: Some("pod_name".to_owned()),
+                uid: Some("pod_uid".to_owned()),
+                ..ObjectMeta::default()
+            }),
+            ..Pod::default()
+        }
+    }
+
+    fn get_metric_value(op_kind: &'static str) -> Option<metrics_runtime::Measurement> {
+        let controller = crate::metrics::CONTROLLER.get().unwrap_or_else(|| {
+            crate::metrics::init().unwrap();
+            crate::metrics::CONTROLLER
+                .get()
+                .expect("failed to init metric container")
+        });
+
+        let key = metrics_core::Key::from_name_and_labels(
+            "k8s_state_ops",
+            vec![metrics_core::Label::new("op_kind", op_kind)],
+        );
+        controller
+            .snapshot()
+            .into_measurements()
+            .into_iter()
+            .find_map(|(candidate_key, measurement)| {
+                if candidate_key == key {
+                    Some(measurement)
+                } else {
+                    None
+                }
+            })
+    }
+
+    fn assert_counter_changed(
+        before: Option<metrics_runtime::Measurement>,
+        after: Option<metrics_runtime::Measurement>,
+        expected_difference: u64,
+    ) {
+        let before = before.unwrap_or_else(|| metrics_runtime::Measurement::Counter(0));
+        let after = after.unwrap_or_else(|| metrics_runtime::Measurement::Counter(0));
+
+        let (before, after) = match (before, after) {
+            (
+                metrics_runtime::Measurement::Counter(before),
+                metrics_runtime::Measurement::Counter(after),
+            ) => (before, after),
+            _ => panic!("metrics kind mismatch"),
+        };
+
+        let difference = after - before;
+
+        assert_eq!(difference, expected_difference);
+    }
+
+    /// Guarantees only one test will run at a time.
+    /// This is required because we assert on a global state, and we don't
+    /// want interference.
+    fn tests_lock() -> MutexGuard<'static, ()> {
+        static INSTANCE: OnceCell<Mutex<()>> = OnceCell::new();
+        INSTANCE.get_or_init(|| Mutex::new(())).lock().unwrap()
+    }
+
+    #[ignore]
```

Comment:
Is there a reason these tests are ignored?

Reply:
I will, thx!

Would be great if we could properly test events.